apple-codesign/src/windows.rs

```diff
@@ -0,0 +1,736 @@
+        let algorithm = self.signature_algorithm()?;
+
+        Ok((self.try_sign(message)?.into(), algorithm))
+    }
+
+    fn key_algorithm(&self) -> Option<KeyAlgorithm> {
+        self.captured.key_algorithm()
+    }
+
+    fn public_key_data(&self) -> Bytes {
+        self.captured.public_key_data()
+    }
+
+    fn signature_algorithm(&self) -> Result<SignatureAlgorithm, X509CertificateError> {
+        self.captured
+            .signature_algorithm()
+            .ok_or(X509CertificateError::UnknownSignatureAlgorithm(format!(
+                "{:?}",
+                self.captured.signature_algorithm_oid()
+            )))
+    }
+
+    fn private_key_data(&self) -> Option<Zeroizing<Vec<u8>>> {
+        None
+    }
+
+    fn rsa_primes(
+        &self,
+    ) -> Result<Option<(Zeroizing<Vec<u8>>, Zeroizing<Vec<u8>>)>, X509CertificateError> {
+        Ok(None)
+    }
+}
+
+impl KeyInfoSigner for StoreCertificate {}
+
+impl PublicKeyPeerDecrypt for StoreCertificate {
+    fn decrypt(&self, ciphertext: &[u8]) -> Result<Vec<u8>, RemoteSignError> {
+        // First, we need to check if the signer has a private key.
+        if self.hkey == NCRYPT_KEY_HANDLE::default() {
+            return Err(RemoteSignError::Crypto(
+                "certificate does not have a private key".into(),
+            ));
+        }
+
+        // We set the OAEP padding info.
+        let padding_info: *mut BCRYPT_OAEP_PADDING_INFO = &mut BCRYPT_OAEP_PADDING_INFO {
+            pszAlgId: BCRYPT_SHA256_ALGORITHM as *mut u16,
+            cbLabel: 0,
+            pbLabel: ptr::null_mut(),
+        };
+
+        // We decrypt using NCryptDecrypt.
+        let mut plaintext: Vec<u8> = Vec::new();
+        let mut plaintext_len: u32 = 0;
+        let result = unsafe {
+            NCryptDecrypt(
+                self.hkey,
+                ciphertext.as_ptr(),
+                ciphertext.len() as u32,
+                padding_info as *mut core::ffi::c_void,
+                ptr::null_mut(),
+                0,
+                &mut plaintext_len,
+                BCRYPT_PAD_OAEP,
+            )
+        };
+        if result != 0 {
+            return Err(RemoteSignError::Crypto(format!("error when attempting to decrypt ciphertext (NCryptDecrypt 1): 0x{:08X}", result)));
+        }
+        plaintext.resize(plaintext_len as usize, 0);
+        let result = unsafe {
+            NCryptDecrypt(
+                self.hkey,
+                ciphertext.as_ptr(),
+                ciphertext.len() as u32,
+                padding_info as *mut core::ffi::c_void,
+                plaintext.as_mut_ptr(),
+                plaintext.len() as u32,
+                &mut plaintext_len,
+                BCRYPT_PAD_OAEP,
+            )
+        };
+        if result != 0 {
+            return Err(RemoteSignError::Crypto(format!("error when attempting to decrypt ciphertext (NCryptDecrypt 2): 0x{:08X}", result)));
+        }
+        plaintext.resize(plaintext_len as usize, 0);
```

Comment:
This looks like duplicate code. Is there a reason it is here?

Reply:
The first call to `NCryptDecrypt` is a size call, meaning we just ask for the size of the plaintext. The second call is where the actual decryption takes place. This is a usual thing with NCrypt API, and I do it also when signing using `NCryptSignHash`.